docs/features/communication/abi_compatible_data_types/index.rst

```diff
@@ -0,0 +1,491 @@
+
+1. **ABI Compatibility**: Processes implemented in different programming languages (C++17 and Rust 1.8x) must interpret a shared memory location consistently as the same native type, provided both have compile-time access to the type definition. This scenario eliminates serialization overhead and allows direct memory access.
+
+2. **Type Description**: Processes responsible for translating data between internal ABI-compatible formats and external serialization formats should perform this translation without compile-time knowledge of type definitions. This runtime interpretation capability is critical for scalability, ensuring that gateway processes and tools for recording/replaying/inspecting data can dynamically handle new data types without recompilation.
+
+
+ABI Compatibility
+-----------------
+
+Our communication feature relies on shared memory to transfer data between processes. For effective zero-copy data exchange, processes written in C++17 and Rust 1.8x must inherently understand the data at shared memory locations identically. Achieving this requires ensuring that data types have consistent, fixed-size memory layouts.
+
+This evaluation initially targets the following process configurations:
+
+* Processes running on the same operating system.
+* Processes running on different operating systems but under the same hypervisor.
+
+Supporting different endianness between processes is explicitly out of scope, as it inherently demands bit manipulation, effectively requiring serialization.
+Different bit widths, however, are implicitly supported by specifying the width of all types and excluding word-size integers.
+
+The following data types shall be supported by the IPC mechanism:
+
+* **Primitive Types**:
+
+  * Boolean
+  * Numeric (fixed-size integers 8-128 bits, signed and unsigned; IEEE 754 floating-point numbers)
+
+* **Sequence Types**:
+
+  * Array (fixed-length)
+
+* **User-Defined Types**:
+
+  * Struct
+  * Tuple
+  * Enum (tag-only)
+  * Variant ("tagged union"; *optional*)
+
+* **Fixed-Size, Variable-Length Containers**:
+
+  * Vector
+  * Queue
+  * Hash map (*optional*)
+  * Hash set (*optional*)
+  * Binary tree (*optional*)
+
+* **Specialized Variants**:
+
+  * Result
+  * Option
+
+All provided data types must ensure fixed size and consistent memory layouts.
+
+Type Description
+----------------
+
+A critical scalability feature involves gateway processes, which subscribe to IPC endpoints and translate ABI-compatible data types into external serialization formats. These gateways require the ability to interpret data without compile-time access to type definitions. To address this, an explicit runtime-readable type description format is necessary. This description allows dynamic, runtime interpretation of data structures, enabling the addition of new IPC topics without recompiling gateway processes.
```

Comment:
See comment above on gateways and their transformations. I think reprocessing is a better driver for this.